pkg/kubecost/allocation.go

```diff
@@ -2131,34 +2131,179 @@ func (asr *AllocationSetRange) NewAccumulation() (*AllocationSet, error) {
 	var err error
+	var as *AllocationSet
+	if clone {
+		as, err = asr.newAccumulation()
+	} else {
+		as, err = asr.accumulate()
+	}
+	if err != nil {
+		return nil, fmt.Errorf("error accumulating all:%s", err)
+	}
+
+	accumulated := NewAllocationSetRange(as)
+	return accumulated, nil
+}
+
+func (asr *AllocationSetRange) accumulateByNone(clone bool) (*AllocationSetRange, error) {
+	if clone {
+		return asr.Clone(), nil
+	}
+	return asr, nil
+}
+func (asr *AllocationSetRange) accumulateByHour(clone bool) (*AllocationSetRange, error) {
+	// ensure that the summary allocation sets have a 1-hour window, if a set exists
+	if len(asr.Allocations) > 0 && asr.Allocations[0].Window.Duration() != time.Hour {
+		return nil, fmt.Errorf("window duration must equal 1 hour; got:%s", asr.Allocations[0].Window.Duration())
+	}
+
+	if clone {
+		return asr.Clone(), nil
+	}
+
+	return asr, nil
+}
+
+func (asr *AllocationSetRange) accumulateByDay(clone bool) (*AllocationSetRange, error) {
+	// if the allocation set window is 1-day, just return the existing allocation set range
+	if len(asr.Allocations) > 0 && asr.Allocations[0].Window.Duration() == time.Hour*24 {
+		return asr, nil
+	}
 
+	var toAccumulate *AllocationSetRange
+	result := NewAllocationSetRange()
 	for i, as := range asr.Allocations {
-		allocSet, err = allocSet.Accumulate(as)
+
+		if as.Window.Duration() != time.Hour {
+			return nil, fmt.Errorf("window duration must equal 1 hour; got:%s", as.Window.Duration())
+		}
+
+		if clone {
+			as = as.Clone()
+		}
+
+		hour := as.Window.Start().Hour()
+
+		if toAccumulate == nil {
+			toAccumulate = NewAllocationSetRange()
+		}
+
+		toAccumulate.Append(as)
+		asAccumulated, err := toAccumulate.accumulate()
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("error accumulating result: %s", err)
+		}
+		toAccumulate = NewAllocationSetRange(asAccumulated)
+
+		if hour == 23 || i == len(asr.Allocations)-1 {
+			if length := len(toAccumulate.Allocations); length != 1 {
+				return nil, fmt.Errorf("failed accumulation, detected %d sets instead of 1", length)
+			}
+			result.Append(toAccumulate.Allocations[0])
+			toAccumulate = nil
 		}
+	}
+	return result, nil
+}
 
-		if allocSet != nil {
+func (asr *AllocationSetRange) accumulateByMonth(clone bool) (*AllocationSetRange, error) {
+	var toAccumulate *AllocationSetRange
+	result := NewAllocationSetRange()
+	for i, as := range asr.Allocations {
+		if as.Window.Duration() != time.Hour*24 {
+			return nil, fmt.Errorf("window duration must equal 24 hours; got:%s", as.Window.Duration())
+		}
 
-			// check if end of asr to sum the final set
-			// If total asr accumulated sum <= resolution return 1 accumulated set
-			if allocSet.Window.Duration() >= resolution || i == len(asr.Allocations)-1 {
-				allocSetRange.Allocations = append(allocSetRange.Allocations, allocSet)
-				allocSet = NewAllocationSet(time.Time{}, time.Time{})
+		if clone {
+			as = as.Clone()
+		}
```

Comment:
This looks like the right idea! Just a little worried about exposing this internal "implementation detail" as a flag to the caller.